lib/spack/spack/cmd/verify.py

```diff
@@ -0,0 +1,93 @@
+
+import spack.store
+import spack.verify
+import spack.environment as ev
+
+description = "Check that all spack packages are on disk as installed"
+section = "admin"
+level = "long"
+
+
+def setup_parser(subparser):
+    setup_parser.parser = subparser
+
+    subparser.add_argument('-l', '--local', action='store_true',
+                           help="Verify only locally installed packages")
+    subparser.add_argument('-j', '--json', action='store_true',
+                           help="Ouptut json-formatted errors")
+    subparser.add_argument('-a', '--all', action='store_true',
+                           help="Verify all packages")
+    subparser.add_argument('files_or_specs', nargs=argparse.REMAINDER,
+                           help="Files or specs to verify")
+
+    type = subparser.add_mutually_exclusive_group()
+    type.add_argument('-s', '--specs', action='store_const', const='specs',
+                      dest='type', default='specs',
+                      help='Treat entries as specs (default)')
+    type.add_argument('-f', '--files', action='store_const', const='files',
+                      dest='type', default='specs',
+                      help="Treat entries as files. Cannot be used with '-a'")
+
+
+def verify(parser, args):
+    local = args.local
+
+    if args.type == 'files':
+        if args.all:
+            setup_parser.parser.print_help()
+            return 1
+
+        for file in args.files_or_specs:
+            results = spack.verify.check_file_manifest(file)
+            if results:
+                if args.json:
+                    print(results.json_string())
+                else:
+                    print(results)
+
+        return 0
+    else:
+        spec_args = spack.cmd.parse_specs(args.files_or_specs)
+
+    if args.all:
+        # setup db query local vs global
+        query = spack.store.db.query_local if local else spack.store.db.query
+
+        # construct spec list
+        if spec_args:
+            spec_list = spack.cmd.parse_specs(args.files_or_specs)
+            specs = []
+            for spec in spec_list:
+                specs += query(spec, installed=True)
+        else:
+            specs = query(installed=True)
+
+    elif args.files_or_specs:
+        # construct disambiguated spec list
+        env = ev.get_env(args, 'verify')
+        specs = list(map(lambda x: spack.cmd.disambiguate_spec(x, env,
+                                                               local=local),
+                         spec_args))
+    else:
+        setup_parser.parser.print_help()
+        return 1
+
+    for spec in specs:
+        tty.debug("Verifying package %s")
+        results = spack.verify.check_spec_manifest(spec)
+        if results:
+            if args.json:
+                print(results.json_string())
+            else:
+                tty.msg("In package %s" % spec.format('{name}/{hash:7}'))
+                print(results)
+        else:
+            tty.debug(results)
```

Comment:
I have to keep scrolling down to the `VerificationResults` class to recall whether `bool(results) == False` means there are errors or not. IMO it should just have a `has_errors` function.

Also I think this should return an integer in keeping with the other logic paths.

Reply:
Spack commands all have an implicit `return 0`.

Reply:
There were two parts 

* Add `has_errors` to `VerificationResults`: I don't see that in the updates
* I assumed that if there were verification errors then this should `return 1` (i.e. that there should be a `return 1` in the `if` portion of this `if/else` statement). If the interpretation of the case where it found verification errors is that the command succeeded (in the sense that it didn't get bad input and was able to complete its verification task) then I could see there is no need for any additional `return` statement.

Reply:
Checked with the requesting user and the tool this would replace does return 1 if it detects errors, although the user was surprised to discover that as he had never checked the return code of the tool.

Addressed the boolean confusion, although I think duck-typing results to `bool` is more pythonic.